### PR TITLE
Maintenance: rework SASL detection

### PR DESCRIPTION
**Pull request by kinkie**
MacOS Homebrew offers Cyrus SASL.

### CODE REVIEW

File: configure.ac
Comment (rousskov) [inline configure.ac:333]: After this PR and other MacOS portability fixes, will Squid _always_ enable LDAP and SASL support on MacOS? If not, is there a way to make this `-Wno-deprecated-declarations` addition _conditional_ as well, so that we do not have to silence useful warnings in other headers on MacOS.

**Review by rousskov — COMMENTED** [review]
I adjusted PR title a little because SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED() updates SQUID_CXXFLAGS rather than CXXFLAGS and because that is a low-level technical detail we can, IMO, omit to emphasize the gravity of these changes. I do not insist on that adjustment.

BTW, I think all these "MacOS compat: X" PR titles should be refactored to become "X on MacOS" (or some such) to avoid unnecessary abbreviations and prefix pollution, but there are too many of them for me to fight :-(.

I also pushed a minor adjustment. Please check.

File: configure.ac
Comment (kinkie) [reply configure.ac:333]: No, the normal behaviours are preserved, e.g. if an administrator decides to not build LDAP helpers, they will not be built.
There are two locations where these headers may be found on MacOS:
- as part of Xcode (where the API we use are marked as deprecated)
- as part of homebrew (where it's the standard openldap  and cyrus-SASL)

With this flag, if the Xcode headers are installed, LDAP squid helpers will build; without it, they won't.

I don't understand the logic on making using this flag conditional: Squid's default behaviour across all systems is to build all helpers that can be built, this flag restores that default behaviour.

At this time, these are the only deprecated APIs we are relying on (checked by grep-ing the output of a build on MacOS for the keyword 'deprecated')

Comment (rousskov) [reply configure.ac:333]: > I don't understand the logic on making using this flag conditional: 

We do not want to disable useful warnings in environments where disabling them does not address any known problems.

> Squid's default behaviour across all systems is to build all helpers that can be built

I do not want to change that default. I also want to preserve/enable LDAP-from-Xcode support where possible. This change request is not about any of that.

If possible, please disable these warnings (i.e. add `-Wno-deprecated-declarations`) only when LDAP helpers are enabled on MacOS.

Comment (kinkie) [reply configure.ac]: done

File: configure.ac
Comment (rousskov) [inline configure.ac:1640]: This call adjusts CXXFLAGS to enable warning-free compilation of code that uses "native" MacOS LDAP headers. This call sits inside SQUID_CHECK_LIB_WORKS(ldap...). I believe this call location is wrong (even if it "works"):

* If adjusted flags are needed for SQUID_CHECK_LDAP_API (or other "Does ldap work?" checks above SQUID_CHECK_LDAP_API) to work correctly, then the flags should be adjusted **earlier**, before executing the code that needs them.
* Otherwise, the adjusted flags are only needed for Squid code that uses LDAP APIs. In that case, they should be adjusted **later**, after SQUID_CHECK_LIB_WORKS() concludes that "ldap works" (i.e. sets LIBLDAP_LIBS) and exits. If this bullet is applicable, the (moved) AS_IF condition will need to test LIBLDAP_LIBS instead of ac_cv_....

N.B. In the above "X needs flags" stands for "X _may_ fail to compile without -Wno-deprecated-declarations". I do understand that we are not checking whether X _will necessarily_ fail without that flag, and I am not requesting that we start checking that.

File: src/auth/basic/SASL/required.m4
Comment (rousskov) [inline src/auth/basic/SASL/required.m4:12]: If both SASL and LDAP are enabled on MacOS, how many `-Wno-deprecated-declarations` flags will be added to CXXFLAGS? If the answer is "more than one", then we need to refactor so that the answer becomes "at most one".

The exact solution will depend on the other change request. For example, if the second bullet in that change request applies, then we probably need to place a single, conditional SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED() call _after_ all LDAP and SASL checks.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: configure.ac
Comment (kinkie) [reply configure.ac:1642]: > This call adjusts SQUID_CXXFLAGS to enable warning-free compilation of code that uses "native" MacOS LDAP headers. This call sits inside SQUID_CHECK_LIB_WORKS(ldap...). I believe this call location is wrong (even if it "works"):

You are right, but not for the reasons you think.
Configure tests pass without the flag. Build doesn't. The reason is that configure doesn't set CXXFLAGS to -Werror; it sets it in SQUID_CXXFLAGS which are not used for the tests but are used for the build. I don't dare touch that, it's likely to be a big can of worms; I've opened draft PR #1702 to facilitate the discussion about this

> * If adjusted flags are needed for SQUID_CHECK_LDAP_API (or other "Does ldap work?" checks above SQUID_CHECK_LDAP_API) to work correctly, then the flags should be adjusted **earlier**, before executing the code that needs them.

After PR 1702, lands (if ever), yes. Right now doing that would result in enabling this flag unconditionally, unless the user were to specify --without-{ldap,sasl} (there are workarounds for this, but it'd be adding more and more complication)

> * Otherwise, the adjusted flags are only needed for Squid code that uses LDAP APIs. In that case, they should be adjusted **later**, after SQUID_CHECK_LIB_WORKS() concludes that "ldap works" (i.e. sets LIBLDAP_LIBS) and exits. If this bullet is applicable, the (moved) AS_IF condition will need to test LIBLDAP_LIBS instead of ac_cv_....

Yes, done this.

> N.B. In the above "X needs flags" stands for "X _may_ fail to compile without -Wno-deprecated-declarations". I do understand that we are not checking whether X _will necessarily_ fail without that flag, and I am not requesting that we start checking that.

Yes

File: src/auth/basic/SASL/required.m4
Comment (kinkie) [reply src/auth/basic/SASL/required.m4]: It wasn't happening in my tests, but I've added code to explicitly avoid duplication

File: configure.ac
Comment (rousskov) [inline configure.ac:1639]: This comment looks misplaced to me.

```suggestion
])
# We can still rely on deprecated APIs on MacOS
```

Personally, I would even lower this comment one more line, placing it inside the AS_IF() statement.

File: configure.ac
Comment (rousskov) [inline configure.ac:1640]: SQUID_CHECK_LIB_WORKS() API documents that the macro sets LIBFOO_LIBS on success. Do we have to test some obscure internal variables instead?!

```suggestion
AS_IF([test "x$squid_host_os" = "xdarwin" -a "x$LIBLDAP_LIBS" != "x"],[
```

**Review by rousskov — DISMISSED** [review]

File: configure.ac
Comment (yadij) [inline configure.ac:1639]: The checks performed to detect a working library involve several compile tests, which could fail. Even if the header ones do not prevent `LIBLDAP_LIBS` being setup and used. The build would break later citing missing includes when they would build fine with the flag added.

I cannot attach this to the correct line. Please add the following `AS_IF` line between the two quoted.
```
    LIBS="$LIBLDAP_PATH $LIBPTHREADS $LIBS"
    AS_IF([test "x$squid_host_os" = "xdarwin"],[CXXFLAGS="-Wno-deprecated-declarations $CXXFLAGS"])
    PKG_CHECK_MODULES([LIBLDAP],[ldap],[],[
```

**Review by yadij — CHANGES_REQUESTED** [review]

File: src/auth/basic/SASL/required.m4
Comment (rousskov) [reply src/auth/basic/SASL/required.m4:13]: > IMO preventing duplicate flags is a larger issue 

I agree that it is a separate issue of sufficient "size" to warrant a dedicated PR.

> that affects a lot of things out of this PR scope

Preventing duplicate flags may affect a lot of future changes. However, it does not affect anything in current official code because we never explicitly call SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED() twice with the same argument AFAICT. This is a (suspected) new problem, (suspected to be) introduced by this PR changes. Moreover, it is possible to refactor this PR to avoid (suspicions about introducing) this problem in the first place.

> I believe the refactor and checking is best done elsewhere, so am against complicating this PR with the request to refactor.

I support (and even encourage) enhancing SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED (or whatever) to avoid flags duplication in a dedicated PR. It is possible to test those enhancements properly without this PR changes (e.g., by temporary introducing duplicated calls), and I hope that Francesco will do that.

I do not support merging this PR until this issue is resolved (one way or another), especially if this PR may result in duplicated flags. Fortunately, there is no rush with landing this PR AFAICT. We can wait for the prerequisite do-not-duplicate-flags PR to land first. Thus, there is currently no conflict in the associated change requests.

File: configure.ac
Comment (kinkie) [reply configure.ac:1639]: removed

Comment (kinkie) [reply configure.ac:1639]: This isn't portable, that flag is currently usupported by clang.
I think what we need to do is:
- when we detect compiler flags early in configure, detect (but not activate) this.
- then here rely on the detected value up there and use it if needed

Comment (kinkie) [reply configure.ac:1639]: This is implemented in commit d46cc2b2dd

Comment (yadij) [reply configure.ac:1639]: FYI, github does not recognize that hash ID so I cannot check the menthioned change.

Comment (rousskov) [reply configure.ac:1639]: > FYI, github does not recognize that hash ID so I cannot check the mentioned change.

Perhaps Francesco forgot to push his changes? IMO, it is usually best to post comments _after_ you push. That way, there is less confusion for the comment reader, and the comment author gets visual feedback if their commit SHAs are incorrect. One can write comments before pushing branch changes but post those comments after the push, but that requires even more discipline/care; IME, GitHub will not render links to SHAs unless an earlier written comment is updated after the push.
 
>  One need to copy the entire long-form hash or use github UI to locate the correct hash manually and copy that one instead.

IME, the situation is much simpler:  The short SHA needs to be known to GitHub (push first!) and long enough for GitHub to show a link to a single commit; seven digits is the current minimum AFAICT. I routinely post my 7-digit short-form SHAs in Squid comments, and they usually work fine. Longer short SHAs will have an even greater chance of success because they are more likely to be unique, of course.

Here is a test with this PR first commit SHA shortened to various lengths:

* six digits (this is the only test line that does not work today): 3b9e56
* seven digits: 3b9e563
* eight digits: 3b9e563a
* sixteen digits: 3b9e563ac88a80b4
* seventeen digits: 3b9e563ac88a80b48
* all forty digits: 3b9e563ac88a80b48e18d92a874ae79c55188328

As Squid repository gets more commits, using eight digits for short SHAs may become routinely necessary to avoid many conflicts. For example, they say that Linux kernel uses at least 11 digits, and git (repository) itself needs at least 10. See also: https://stackoverflow.com/questions/18134627/how-much-of-a-git-sha-is-generally-considered-necessary-to-uniquely-identify-a

**Comment by yadij** [comment]
@kinkie , ping. PR #1736 (was PR #1710) has merged. I have merged and resolved the resulting conflicts. Over to you for the remaining work on this.

**Comment by kinkie** [comment]
With the reworked SASL detection and @yadij 's changes to LDAP, MacOS/homebrew builds
with these environment variables at ./configure time:

PKG_CONFIG_PATH="/opt/homebrew/opt/openldap/lib/pkgconfig:/opt/homebrew/opt/cyrus-sasl/lib/pkgconfig" CPPFLAGS=-I/opt/homebrew/include

CPPFLAGS can hopefully be removed once openssl header detection is improved

provided that the following homebrew packages are intalled:
autoconf-archive
automake
pkg-config
coreutils
cppunit
cyrus-sasl
gawk
gnutls
libtool
make
nettle
openssl@3

and
Xcode with relative command line utilities

File: src/auth/basic/SASL/required.m4
Comment (rousskov) [reply src/auth/basic/SASL/required.m4]: PR branch commit 0d9e75fc8129d29f6e534c5801a0de2060910f67 removed flags/warning manipulation, obsoleting this change request.

Comment (kinkie) [reply src/auth/basic/SASL/required.m4]: That's correct. In the end @yadij 's rework of lib detection simplified things tremendously and
uncovered the actual problem - an overcomplicated SASL detection.

File: src/acl/external/kerberos_ldap_group/required.m4
Comment (yadij) [inline src/acl/external/kerberos_ldap_group/required.m4:9]: This does need the LDAP check 
```suggestion
  AS_IF([test  "x$LIBLDAP_LIBS" != "x" -a "x$LIBSASL_LIBS" != "x"],[
```

File: src/auth/basic/SASL/required.m4
Comment (yadij) [inline src/auth/basic/SASL/required.m4:10]: keeping the single line would be better
```suggestion
AS_IF([test "x$LIBSASL_LIBS" != "x"],[BUILD_HELPER="SASL"])
```

File: configure.ac
Comment (yadij) [inline configure.ac:1210]: Parameter two is the library name/description for use in the `./configure --help` text.
Parameter 3 must be the same variable passed to `PKG_CHECK_MODULES()` parameter 1.
```suggestion
SQUID_AUTO_LIB(sasl,[Cyrus SASL],[LIBSASL])
```

**Review by yadij — CHANGES_REQUESTED** [review]
Please add missing release notes about the new `./configure --without-sasl` option.

**Comment by kinkie** [comment]
Thanks @yadij 

Excerpt from ./configure output on MacOS Sonoma:
```
checking for LIBSASL... yes
checking for sasl/sasl.h... yes
checking for sasl.h... no
configure: Library 'sasl' support: yes (auto)  -L/opt/homebrew/Cellar/cyrus-sasl/2.1.28_2/lib -lsasl2
[...]
configure: auth/basic helpers to be built:  DB LDAP NCSA PAM POP3 RADIUS SASL SMB fake getpwnam
```

File: doc/release-notes/release-7.sgml.in
Comment (yadij) [inline doc/release-notes/release-7.sgml.in:198]:
```suggestion
	<p>Disable auto-detection of Cyrus SASL (or compatible) library.
```

File: acinclude/squid-util.m4
Comment (yadij) [inline acinclude/squid-util.m4:347]: This `AC_DEFINE(HAVE_SASL_DARWIN` being removed should be matched with removal of all uses of the `HAVE_SASL_DARWIN` in the code. Most look trivial and already pointless.

Though there is one `#elif HAVE_SASL_DARWIN` in `src/acl/external/kerberos_ldap_group/support_sasl.cc` which will need a closer look and a test on MacOS that the helper still works after this PR.

**Review by yadij — CHANGES_REQUESTED** [review]
Looks good. One last issue with `HAVE_SASL_DARWIN`, sorry for not noticing earlier.

File: acinclude/squid-util.m4
Comment (kinkie) [reply acinclude/squid-util.m4:347]: I can check it builds, I can't runtime test it, but given that after this we're relying on Cyrus SASL and not on the Mac one, it should work just like everywhere else provided that it builds

Comment (kinkie) [reply acinclude/squid-util.m4:347]: Confirmed building

**Review by yadij — APPROVED** [review]

**Pull request closed without merging**
